### PR TITLE
plan, cmd: plan, cmd: improve the code readability of planner

### PR DESCRIPTION
**Pull request by zz-jason**
Main Modifications:
- rename and tiny refactor the function `getPermutation` to `getMaxSortPrefix`
- rename the code files which contain the implementation of logical optimization rule to the `rule_` prefixed one
- tiny refactor `LogicalAggregation.getStreamAggs()`
- tiny refactor `DataSource.convertToIndexScan()`, which further causes the modification of `explain-test`

### CODE REVIEW

File: cmd/explaintest/r/explain_easy_stats.result [hunk 1]
```diff
@@ -100,20 +100,20 @@ explain select c2 = (select c2 from t2 where t1.c1 = t2.c1 order by c1 limit 1)
 id	parents	children	task	operator info	count
 TableScan_15			cop	table:t1, range:[-inf,+inf], keep order:false	1999.00
 TableReader_16	Apply_14		root	data:TableScan_15	1999.00
-IndexScan_33	Selection_35		cop	table:t2, index:c1, range:[<nil>,+inf], keep order:true	1.25
-Selection_35	Limit_36	IndexScan_33	cop	eq(test.t1.c1, test.t2.c1)	1.00
-Limit_36		Selection_35	cop	offset:0, count:1	1.00
-TableScan_34			cop	table:t2, keep order:false	1.00
-IndexLookUp_37	Limit_21		root	index:Limit_36, table:TableScan_34	1.00
-Limit_21	Apply_14	IndexLookUp_37	root	offset:0, count:1	1.00
+IndexScan_32	Selection_34		cop	table:t2, index:c1, range:[<nil>,+inf], keep order:true	1.25
```
Comment (shenli) [inline cmd/explaintest/r/explain_easy_stats.result]: Why is the ID changed?

File: plan/exhaust_physical_plans.go [hunk 2]
```diff
@@ -541,33 +545,38 @@ func (la *LogicalAggregation) getStreamAggs(prop *requiredProp) []PhysicalPlan {
...
+			continue
+		}
+
+		childProp.cols = possibleChildProperty[:len(sortColOffsets)]
+		if !prop.isPrefix(childProp) {
 			continue
 		}
-		for _, tp := range wholeTaskTypes {
+
+		for _, taskTp := range wholeTaskTypes {
```
Comment (shenli) [inline plan/exhaust_physical_plans.go]: Can we define a new constant slice that contains copSingleReadTaskType and rootTaskType? So we could save the following if.

File: cmd/explaintest/r/explain_easy_stats.result (hunk 1 above)
Comment (zz-jason) [reply cmd/explaintest/r/explain_easy_stats.result]: due to the refactor of `DataSource.convertToIndexScan()`

Comment (shenli) [reply cmd/explaintest/r/explain_easy_stats.result]: Got it.

File: plan/exhaust_physical_plans.go [hunk 3]
```diff
@@ -541,33 +545,38 @@ func (la *LogicalAggregation) getStreamAggs(prop *requiredProp) []PhysicalPlan {
...
-			if childProp.expectedCnt < prop.expectedCnt {
-				childProp.expectedCnt = prop.expectedCnt
-			}
-			if !prop.isPrefix(childProp) {
+			if taskTp == copDoubleReadTaskType {
 				continue
 			}
+
+			copiedChildProperty := new(requiredProp)
+			*copiedChildProperty = *childProp // It's ok to not deep copy the "cols" field.
```
Comment (coocood) [inline plan/exhaust_physical_plans.go]: Why define a child property outside of the loop and use copy?
for performance?

**Comment by coocood** [comment]
LGTM

File: plan/exhaust_physical_plans.go [hunk 4]
```diff
@@ -541,33 +545,33 @@ func (la *LogicalAggregation) getStreamAggs(prop *requiredProp) []PhysicalPlan {
...
-			if !prop.isPrefix(childProp) {
-				continue
-			}
+
+		childProp.cols = possibleChildProperty[:len(sortColOffsets)]
+		if !prop.isPrefix(childProp) {
+			continue
+		}
+
+		for _, taskTp := range []taskType{copSingleReadTaskType, rootTaskType} {
```
Comment (shenli) [inline plan/exhaust_physical_plans.go]: Add a comment such as:
" DoubleRead can't meet the stream aggregation's require prop. So we do not consider it here."

**Comment by zz-jason** [comment]
@shenli PTAL

**Comment by shenli** [comment]
LGTM

**Pull request merged**